### ISSUE TITLE
Very stupid bug on FILE_INFO::verify_file_certs()

### ISSUE THREAD

**Issue by romw** [BOINC/boinc-ng#1141]
**Reported by matszpk on 13 Mar 42001188 11:20 UTC**
```
while (dir_scan(file, dir, sizeof(file))) {
```

should be replaced by:

```
while (!dir_scan(file, dir, sizeof(file))) {
```

dir_scan returns zero if successfully reads next dir entry.

**Comment by romw** [comment]
**Commented by smoe on 22 Mar 42673302 00:16 UTC**
I can confirm the issue to be still present with 7.0.34 and to effect also lib/crypt.cpp. The patch below (compiles here) presumably fixes both. When e.g. applied on empty directories, like while changing projects, possibly, I could imagine the code to have side effects that might contribute to https://boinc.berkeley.edu/trac/ticket/1203 . Just a hunch.

Index: boinc/client/cs_files.cpp

===================================================================[boinc.orig/client/cs_files.cpp 2012-08-30 21:03:35.490956131 +0200[[BR]([BR]]---)]+++ boinc/client/cs_files.cpp 2012-09-02 10:40:36.478032727 +0200[-90,12 +90,12 @@[[BR]([BR]]@@)]// Is app signed by one of the Application Certifiers?[FILE_INFO::verify_file_certs() {[[BR]([BR]]//[[BR]]bool)]- char file![char file[MAXPATHLEN](256];[[BR]]+);[bool retval = false;[[BR]([BR]])][if (!is_dir(CERTIFICATE_DIRECTORY)) return false;[[BR]([BR]])] DIRREF dir = dir_open(CERTIFICATE_DIRECTORY);[while (dir_scan(file, dir, sizeof(file))) {[[BR]([BR]]-)]+ while (!dir_scan(file, dir, sizeof(file))) {[if (cert_verify_file(cert_sigs, file, CERTIFICATE_DIRECTORY)) {[[BR]([BR]])] msg_printf(project, MSG_INFO,["Signature verified using certificate %s", file[[BR]([BR]])]@@ -156,7 +156,7 @@[FILE_INFO::verify_file([[BR]([BR]]int)] bool verify_contents, bool show_errors, bool allow_async[{[[BR]([BR]]))]- char cksum![pathname![256](64],);[char cksum![64]([BR]]+), pathname[bool verified;[[BR](MAXPATHLEN];[[BR]])] int retval;[double size, local_nbytes;[[BR]([BR]])]@@ -177,7 +177,9 @@[//[[BR]([BR]])] if (download_gzipped && !boinc_file_exists(pathname)) {[char gzpath[MAXPATHLEN]([BR]]);[sprintf(gzpath, "%s.gz", pathname);[[BR]([BR]]-)]+ snprintf(gzpath, sizeof(gzpath), "%s.gz", pathname);[//FIXME: a distinction is missing for the case that the .gz suffix goes beyond the MAXPATHLEN and is hence not found because[[BR]([BR]]+)]+ // one should then not reperform the download as intended below[if (boinc_file_exists(gzpath) ) {[[BR]([BR]])] if (allow_async && nbytes > ASYNC_FILE_THRESHOLD) {[ASYNC_VERIFY* avp = new ASYNC_VERIFY;[[BR]([BR]])]Index: boinc/lib/crypt.cpp[boinc.orig/lib/crypt.cpp 2012-09-02 10:27:39.420339596 +0200[[BR]([BR]]===================================================================[[BR]]---)]+++ boinc/lib/crypt.cpp 2012-09-02 10:42:50.568254414 +0200[-604,7 +604,7 @@[[BR]([BR]]@@)] DIRREF dir = dir_open(certPath);[char file[MAXPATHLEN]([BR]][[BR]]);[while (dir_scan(file, dir, sizeof(file))) {[[BR]([BR]]-)]+ while (!dir_scan(file, dir, sizeof(file))) {[char fpath[MAXPATHLEN]([BR]]);[snprintf(fpath, sizeof(fpath), "%s/%s", certPath, file);[[BR]([BR]])] // TODO : replace '128'

**Comment by romw** [comment]
**Commented by davea on 10 Jul 42674542 17:21 UTC**
(In [26071]) - client: fix bug that broke file signing with X.509 certificates.
    From matszpk.  Fixes #1168.

**Comment by romw** [comment]
**Commented by romw on 6 Feb 42701781 18:53 UTC**
(In [26094])     - client: fix bug that broke file signing with X.509 certificates.
        From matszpk.  Fixes #1168.
